criu/memfd.c

```diff
@@ -335,18 +335,23 @@ int memfd_open(struct file_desc *d, u32 *fdflags)
 
 	/* Reopen the fd with original permissions */
 	flags = fdflags ? *fdflags : mfe->flags;
-	/*
-	 * Ideally we should call compat version open() to not force the
-	 * O_LARGEFILE file flag with regular open(). It doesn't seem that
-	 * important though.
-	 */
-	_fd = __open_proc(PROC_SELF, 0, flags, "fd/%d", fd);
-	if (_fd < 0) {
-		pr_perror("Can't reopen memfd id=%d", mfe->id);
+	if ((flags & O_ACCMODE) != O_RDWR) {
+		/*
+		 * Ideally we should call compat version open() to not force the
+		 * O_LARGEFILE file flag with regular open(). It doesn't seem that
+		 * important though.
+		 */
+		_fd = __open_proc(PROC_SELF, 0, flags, "fd/%d", fd);
+		if (_fd < 0) {
+			pr_perror("Can't reopen memfd id=%d", mfe->id);
+			goto err;
+		}
+		close(fd);
+		fd = _fd;
+	} else if (fcntl(fd, F_SETFL, flags)) {
```

Comment:
We can do this only once for each inode.  If this memfd_open is called more than one time for one memfd, it means that we have two non-shared file descriptors.

Reply:
Done.